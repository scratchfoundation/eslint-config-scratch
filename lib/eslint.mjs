import eslintConfigPrettier from 'eslint-config-prettier/flat'
import formatjs from 'eslint-plugin-formatjs'
import html from 'eslint-plugin-html'
import htmlSettings from 'eslint-plugin-html/src/settings.js'
import importPlugin from 'eslint-plugin-import'
import jsdoc from 'eslint-plugin-jsdoc'
import jsxA11y from 'eslint-plugin-jsx-a11y'
import markdown from 'eslint-plugin-markdown'
import react from 'eslint-plugin-react'
import reactHooks from 'eslint-plugin-react-hooks'
import globals from 'globals'
import tseslint from 'typescript-eslint'
import eslintComments from '@eslint-community/eslint-plugin-eslint-comments/configs'
import eslint from '@eslint/js'

// See https://www.npmjs.com/package/eslint-plugin-html#user-content-settings
const htmlExtensions = htmlSettings.getSettings({}).htmlExtensions

// '.html' => '**/*.html'
const htmlGlobs = htmlExtensions.map(ext => `**/*${ext}`)

/**
 * @typedef {import('eslint').Linter.Globals} Globals
 * @typedef {keyof globals} GlobalsKey
 * @typedef {Globals | GlobalsKey} GlobalsObjOrKey
 */

/**
 * Flatten the globals passed to `makeScratchConfig` into an object suitable for ESLint's `globals` option.
 * @param {GlobalsObjOrKey | GlobalsObjOrKey[]} [globalsIn] The globals to flatten.
 * @returns {Globals|undefined} Flattened globals object for ESLint.
 */
const flattenGlobals = globalsIn => {
  if (!globalsIn) return

  /**
   *
   * @param {Globals} globalsAcc Globals accumulator.
   * @param {GlobalsObjOrKey} objOrKey A globals object or key to add to the accumulator.
   * @returns {Globals} The accumulator after adding the current globals object or key.
   */
  const globalsReducer = (globalsAcc, objOrKey) => {
    if (typeof objOrKey === 'string') {
      const globalsForKey = globals[objOrKey]
      if (!globalsForKey) {
        throw new Error(`Invalid globals name. Not a key from the globals package: ${objOrKey}`)
      }
      Object.assign(globalsAcc, globalsForKey)
    } else {
      Object.assign(globalsAcc, objOrKey)
    }

    return globalsAcc
  }

  if (Array.isArray(globalsIn)) {
    return globalsIn.reduce(globalsReducer, {})
  }

  return globalsReducer({}, globalsIn)
}

/**
 * Create an ESLint configuration for Scratch style.
 * Supports JavaScript, TypeScript, and React (JSX/TSX) files.
 * Setting `tsconfigRootDir` enables type-aware rules, some of which apply even in JavaScript files.
 * @param {object} options Configuration options
 * @param {string} [options.tsconfigRootDir] Enable type checking by setting the root TypeScript config directory.
 * @param {GlobalsObjOrKey | GlobalsObjOrKey[]} [options.globals] Globals to provide to ESLint.
 * This can be expressed as:
 * - a single string, such as `'browser'`, corresponding to a key in the `globals` package.
 * - a single object as described in the "Specifying Globals" section of the ESLint documentation:
 *   https://eslint.org/docs/latest/use/configure/language-options#using-configuration-files
 * - an array of zero or more elements, each of which can be either of the above
 * @example
 * // eslint.config.mjs
 * export default makeScratchConfig({tsconfigRootDir: import.meta.dirname, globals: 'node'})
 * @example
 * // eslint.config.mjs
 * export default [
 *   ...makeScratchConfig({tsconfigRootDir: import.meta.dirname, globals: 'browser'}),
 *   {
 *     // customization
 *   }
 * ]
 * @returns {import('typescript-eslint').ConfigArray} An ESLint configuration array.
 */
const makeEslintConfig = ({ tsconfigRootDir, globals: globalsIn } = {}) => {
  const flattenedGlobals = flattenGlobals(globalsIn)

  return tseslint.config(
    // Start with recommended rules from ESLint and TypeScript ESLint.
    {
      extends: [
        eslint.configs.recommended,
        tsconfigRootDir ? tseslint.configs.recommendedTypeChecked : tseslint.configs.recommended,
        tsconfigRootDir ? tseslint.configs.stylisticTypeChecked : tseslint.configs.stylistic,
      ],
      languageOptions: {
        parserOptions: {
          ...(tsconfigRootDir
            ? {
                projectService: true,
                tsconfigRootDir,
              }
            : {}),
        },
        ...(globalsIn ? { globals: flattenedGlobals } : {}),
      },
    },
    // eslint-plugin-formatjs
    {
      plugins: {
        formatjs,
      },
      rules: {
        'formatjs/no-offset': ['error'],
      },
    },
    // eslint-plugin-html
    {
      files: htmlGlobs,
      plugins: { html },
      settings: {
        'html/html-extensions': htmlExtensions,
      },
    },
    // eslint-plugin-import
    {
      plugins: importPlugin.flatConfigs.recommended.plugins,
      rules: {
        'import/no-duplicates': 'error', // Forbid duplicate imports
      },
    },
    // eslint-plugin-jsdoc
    jsdoc.configs['flat/recommended-error'],
    {
      files: ['**/*.ts', '**/*.tsx', '**/*.mts', '**/*.cts'],
      extends: [jsdoc.configs['flat/recommended-typescript-error']],
    },
    {
      rules: {
        // If JSDoc comments are present, they must be informative (non-trivial).
        // For example, the description "The foo." on a variable called "foo" is not informative.
        // https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/informative-docs.md
        'jsdoc/informative-docs': ['error'],

        // Require JSDoc comments on public / exported items.
        // https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/require-jsdoc.md
        'jsdoc/require-jsdoc': [
          'error',
          {
            publicOnly: true,
          },
        ],
      },
    },
    // eslint-plugin-jsx-a11y
    jsxA11y.flatConfigs.recommended,
    // eslint-plugin-markdown
    markdown.configs.recommended,
    // eslint-plugin-react
    {
      plugins: {
        react,
      },
      rules: {
        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-danger.md
        'react/no-danger': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/self-closing-comp.md
        'react/self-closing-comp': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-boolean-value.md
        'react/jsx-boolean-value': ['error', 'never'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-closing-bracket-location.md
        'react/jsx-closing-bracket-location': ['error', 'line-aligned'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-curly-spacing.md
        'react/jsx-curly-spacing': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-equals-spacing.md
        'react/jsx-equals-spacing': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md
        'react/jsx-filename-extension': ['error', { extensions: ['.jsx', '.tsx'] }],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-first-prop-new-line.md
        'react/jsx-first-prop-new-line': ['error', 'multiline'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-handler-names.md
        'react/jsx-handler-names': ['error', { checkLocalVariables: true, eventHandlerPrefix: false }],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-indent.md
        'react/jsx-indent': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md
        'react/jsx-no-bind': ['error', { ignoreRefs: true, allowArrowFunctions: true }],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-pascal-case.md
        'react/jsx-pascal-case': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-tag-spacing.md
        'react/jsx-tag-spacing': ['error'],

        // https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-wrap-multilines.md
        'react/jsx-wrap-multilines': ['error'],
      },
      settings: {
        react: {
          version: 'detect',
        },
      },
    },
    // eslint-plugin-react-hooks
    {
      extends: [reactHooks.configs['recommended-latest']],
      rules: {
        // https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/README.md#advanced-configuration
        'react-hooks/exhaustive-deps': ['error', { additionalHooks: '^useAsync$' }],
      },
    },
    // typescript-eslint
    {
      rules: {
        // https://typescript-eslint.io/rules/no-non-null-asserted-nullish-coalescing/
        '@typescript-eslint/no-non-null-asserted-nullish-coalescing': ['error'],

        // https://typescript-eslint.io/rules/no-useless-constructor/
        '@typescript-eslint/no-useless-constructor': ['error'],

        // https://typescript-eslint.io/rules/no-non-null-assertion
        '@typescript-eslint/no-non-null-assertion': ['error'],

        // Rules that require type information
        ...(tsconfigRootDir
          ? {
              // https://typescript-eslint.io/rules/no-unnecessary-condition/
              '@typescript-eslint/no-unnecessary-condition': ['error'],

              // https://typescript-eslint.io/rules/require-await/
              '@typescript-eslint/require-await': ['error'],
            }
          : {}),
      },
    },
    // @eslint-community/eslint-plugin-eslint-comments
    {
      extends: [
        // @ts-expect-error This plugin's recommended rules don't quite match the type `tseslint.config` expects.
        eslintComments.recommended,
      ],
      rules: {
        // require a description for eslint control comments other than `eslint-enable`
        '@eslint-community/eslint-comments/require-description': ['error', { ignore: ['eslint-enable'] }],
      },
    },
    // @eslint/js
    {
      rules: {
        // https://eslint.org/docs/latest/rules/arrow-body-style
        'arrow-body-style': ['error', 'as-needed'],

        // https://eslint.org/docs/latest/rules/no-duplicate-imports
        'no-duplicate-imports': ['error'],

        // https://eslint.org/docs/latest/rules/no-template-curly-in-string
        'no-template-curly-in-string': ['error'],

        // https://eslint.org/docs/latest/rules/no-useless-computed-key
        'no-useless-computed-key': ['error'],

        // https://eslint.org/docs/latest/rules/no-useless-rename
        'no-useless-rename': ['error'],

        // https://eslint.org/docs/latest/rules/prefer-arrow-callback
        'prefer-arrow-callback': ['error'],

        // https://eslint.org/docs/latest/rules/prefer-const#destructuring
        'prefer-const': ['error'],

        // https://eslint.org/docs/latest/rules/prefer-spread
        'prefer-spread': ['error'],

        // https://eslint.org/docs/latest/rules/require-atomic-updates
        'require-atomic-updates': ['error'],

        // https://eslint.org/docs/latest/rules/symbol-description
        'symbol-description': ['error'],
      },
    },
    // Keep `eslintConfigPrettier` last to turn off rules that conflict with Prettier
    eslintConfigPrettier,
  )
}

export { makeEslintConfig }
