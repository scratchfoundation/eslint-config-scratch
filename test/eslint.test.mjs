import { ESLint } from 'eslint'
import path from 'path'
import util from 'util'
import { beforeAll, describe, expect, test } from 'vitest'

/**
 * @typedef {object} EslintTestInfo
 * @property {string} name - the title/message for this test
 * @property {string} filePath - the path to the file to lint
 * @property {number} warningCount - the number of warnings to expect
 * @property {number} errorCount - the number of errors to expect
 */

// TSX is omitted because TypeScript insists on fully knowing the React types,
// and I would rather not add React as a dependency of eslint-config-scratch.
/** @type {Record<string, EslintTestInfo[]>} */
const testInfo = {
  recommended: [
    {
      name: 'Plain JS (good)',
      filePath: 'plain.good.mjs',
      warningCount: 0,
      errorCount: 0,
    },
    {
      name: 'React JSX (good)',
      filePath: 'react.good.jsx',
      warningCount: 0,
      errorCount: 0,
    },
    {
      name: 'Plain JS (bad)',
      filePath: 'plain.bad.mjs',
      warningCount: 0,
      errorCount: 3,
    },
    {
      name: 'React JSX (bad)',
      filePath: 'react.bad.jsx',
      warningCount: 0,
      errorCount: 2,
    },
    {
      name: 'Plain TS (good)',
      filePath: 'plain.good.ts',
      warningCount: 0,
      errorCount: 0,
    },
    {
      name: 'Plain TS (bad)',
      filePath: 'plain.bad.ts',
      warningCount: 0,
      errorCount: 5,
    },
  ],
  legacy: [
    {
      name: 'Plain JS (good)',
      filePath: 'plain.good.mjs',
      warningCount: 0,
      errorCount: 0,
    },
    {
      name: 'React JSX (good)',
      filePath: 'react.good.jsx',
      warningCount: 0,
      errorCount: 0,
    },
    {
      name: 'Plain JS (bad)',
      filePath: 'plain.bad.mjs',
      warningCount: 0,
      errorCount: 4,
    },
    {
      name: 'React JSX (bad)',
      filePath: 'react.bad.jsx',
      warningCount: 0,
      errorCount: 2,
    },
    {
      name: 'Plain TS (good)',
      filePath: 'plain.good.ts',
      warningCount: 0,
      errorCount: 0,
    },
    {
      name: 'Plain TS (bad)',
      filePath: 'plain.bad.ts',
      warningCount: 0,
      errorCount: 5,
    },
  ],
}

/**
 * Create a snapshot of a lint message.
 * Excludes properties that may change without affecting correctness, such as human-readable text.
 * @param {ESLint.LintMessage} result - the lint message to filter
 * @returns {object} a filtered snapshot of the lint message
 */
const messageSnapshot = result =>
  Object.fromEntries(
    Object.entries(result).filter(([k]) =>
      ['line', 'column', 'endLine', 'endColumn', 'messageId', 'nodeType', 'ruleId'].includes(k),
    ),
  )

test('make sure eslint works at all', async () => {
  const source = 'foo(42)'
  const eslint = new ESLint({
    overrideConfigFile: true,
  })
  const results = await eslint.lintText(source)
  expect(results).toBeDefined()
  expect(results.length).toBe(1)
  expect(results[0].warningCount).toEqual(0)
  expect(results[0].errorCount).toEqual(0)
})

describe.concurrent.for(Object.entries(testInfo))('$0', ([subdir, testList]) => {
  /**
   * @type {ESLint.LintResult[]}
   */
  let results

  // Linting one file at a time takes much longer
  beforeAll(async () => {
    const eslint = new ESLint({
      overrideConfigFile: path.resolve(import.meta.dirname, subdir, 'eslint.config.mjs'),
    })
    results = await eslint.lintFiles(testList.map(info => path.resolve(import.meta.dirname, subdir, info.filePath)))
  })

  test('Results container', () => {
    expect(results).toBeDefined()
    expect(results.length).toBe(testList.length)
  })

  testList.forEach(({ name, filePath, warningCount, errorCount }, i) => {
    test(name, () => {
      expect(path.resolve(results[i].filePath)).toBe(path.resolve(import.meta.dirname, subdir, filePath))
      expect(results[i].warningCount, util.inspect(results[i])).toBe(warningCount)
      expect(results[i].errorCount, util.inspect(results[i])).toBe(errorCount)
      expect(results[i].messages.map(messageSnapshot), util.inspect(results[i].messages)).toMatchSnapshot()
    })
  })
})
