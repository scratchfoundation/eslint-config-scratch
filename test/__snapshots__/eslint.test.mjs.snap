// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'recommended' > Plain TS (bad) 1`] = `
[
  {
    "column": 7,
    "endColumn": 25,
    "endLine": 5,
    "line": 5,
    "messageId": "noInferrableType",
    "nodeType": "VariableDeclarator",
    "ruleId": "@typescript-eslint/no-inferrable-types",
  },
  {
    "column": 7,
    "endColumn": 12,
    "endLine": 5,
    "line": 5,
    "messageId": "unusedVar",
    "nodeType": null,
    "ruleId": "@typescript-eslint/no-unused-vars",
  },
  {
    "column": 10,
    "endColumn": 13,
    "endLine": 8,
    "line": 8,
    "messageId": "unusedVar",
    "nodeType": null,
    "ruleId": "@typescript-eslint/no-unused-vars",
  },
  {
    "column": 14,
    "endColumn": 26,
    "endLine": 17,
    "line": 17,
    "messageId": "anyAssignment",
    "nodeType": "VariableDeclarator",
    "ruleId": "@typescript-eslint/no-unsafe-assignment",
  },
  {
    "column": 20,
    "endColumn": 24,
    "endLine": 17,
    "line": 17,
    "messageId": "unsafeCall",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-unsafe-call",
  },
]
`;

exports[`'recommended' > Plain TS (good) 1`] = `[]`;

exports[`'recommended' > React JSX (bad) 1`] = `
[
  {
    "column": 10,
    "endColumn": 13,
    "endLine": 4,
    "line": 4,
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "no-unused-vars",
  },
  {
    "column": 20,
    "endColumn": 24,
    "endLine": 14,
    "line": 14,
    "messageId": "undef",
    "nodeType": "Identifier",
    "ruleId": "no-undef",
  },
]
`;

exports[`'recommended' > React JSX (good) 1`] = `[]`;

exports[`'recommended' > plain JS (bad) 1`] = `
[
  {
    "column": 10,
    "endColumn": 13,
    "endLine": 4,
    "line": 4,
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "no-unused-vars",
  },
  {
    "column": 26,
    "endColumn": 46,
    "endLine": 12,
    "line": 12,
    "messageId": "noJSXWithExtension",
    "nodeType": "JSXElement",
    "ruleId": "react/jsx-filename-extension",
  },
  {
    "column": 20,
    "endColumn": 24,
    "endLine": 15,
    "line": 15,
    "messageId": "undef",
    "nodeType": "Identifier",
    "ruleId": "no-undef",
  },
]
`;

exports[`'recommended' > plain JS (good) 1`] = `[]`;
